// rainbow-node-sdk/lib/config/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
let conf = {
    sandbox: {
        http: {
            host: "sandbox.openrainbow.com",
            port: "443",
            protocol: "https"
        },
        xmpp: {
            host: "sandbox.openrainbow.com",
            port: "443",
            protocol: "wss",
            timeBetweenXmppRequests: "20"
        }
    },
    official: {
        http: {
            host: "openrainbow.com",
            port: "443",
            protocol: "https"
        },
        xmpp: {
            host: "openrainbow.com",
            port: "443",
            protocol: "wss",
            timeBetweenXmppRequests: "20"
        }
    },
    any: {
        http: {
            host: "",
            port: "443",
            protocol: "https"
        },
        xmpp: {
            host: "",
            port: "443",
            protocol: "wss",
            timeBetweenXmppRequests: "20"
        }
    },
    logs: {
        path: "/var/tmp/rainbowsdk/",
        level: "info",
        color: false,
        enableConsoleLog: true,
        "system-dev": {
            internals: false,
            http: false
        },
        zippedArchive: true,
        maxSize: "10m",
        maxFiles: null
    },
    im: {
        sendReadReceipt: true,
        messageMaxLength: 1024,
        sendMessageToConnectedUser: false,
        conversationsRetrievedFormat: "small",
        storeMessages: false,
        /* copyMessage: true, /* https://xmpp.org/extensions/xep-0334.html#hints :
        No copies :
            Messages with the <no-copy/> hint should not be copied to addresses other than the one to which it is addressed, for example through Message Carbons (XEP-0280) [3].
            This hint MUST only be included on messages addressed to full JIDs and explicitly does not override the behaviour defined in XMPP IM [1] for handling messages to bare JIDs,
            which may involve copying to multiple resources, or multiple occupants in a Multi-User Chat (XEP-0045) [6] room.
        // */
        nbMaxConversations: 15,
        rateLimitPerHour: 100
    },
    mode: "xmpp",
    debug: true,
    permitSearchFromPhoneBook: true,
    displayOrder: "firstLast",
    servicesToStart: {
        presence: {
            start_up: true,
            optional: false
        },
        contacts: {
            start_up: true,
            optional: false
        },
        conversations: {
            start_up: true,
            optional: false
        },
        im: {
            start_up: true,
            optional: false
        },
        profiles: {
            start_up: true,
            optional: false
        },
        groups: {
            start_up: true,
            optional: false
        },
        bubbles: {
            start_up: true,
            optional: true
        },
        telephony: {
            start_up: true,
            optional: true
        },
        channels: {
            start_up: true,
            optional: true
        },
        admin: {
            start_up: true,
            optional: true
        },
        fileServer: {
            start_up: true,
            optional: true
        },
        fileStorage: {
            start_up: true,
            optional: true
        },
        calllog: {
            start_up: true,
            optional: true
        },
        favorites: {
            start_up: true,
            optional: true
        },
        invitation: {
            start_up: true,
            optional: true
        },
        settings: {
            start_up: true,
            optional: true
        } //need services : ( XMPPService, _rest : RESTService)
    }
};
exports.conf = conf;
module.exports = conf;
